code/game/machinery/doppler_array.dm

```diff
@@ -11,24 +11,51 @@ GLOBAL_LIST_EMPTY(doppler_arrays)
 	var/integrated = FALSE
 	var/max_dist = 150
 	verb_say = "states coldly"
+	var/list/message_log = list()
+	var/output_log
 
 /obj/machinery/doppler_array/Initialize()
 	. = ..()
 	GLOB.doppler_arrays += src
 
 /obj/machinery/doppler_array/ComponentInitialize()
+	. = ..()
 	AddComponent(/datum/component/simple_rotation,ROTATION_ALTCLICK | ROTATION_CLOCKWISE,null,null,CALLBACK(src,.proc/rot_message))
 
 /obj/machinery/doppler_array/Destroy()
 	GLOB.doppler_arrays -= src
 	return ..()
 
-/obj/machinery/doppler_array/examine(mob/user)
-	..()
-	to_chat(user, "<span class='notice'>Its dish is facing to the [dir2text(dir)].</span>")
 
-/obj/machinery/doppler_array/process()
-	return PROCESS_KILL
+/obj/machinery/doppler_array/ui_interact(mob/user)
+	. = ..()
+	if(stat)
+		return FALSE
+
+	var/dat = {"<html><head><title>[src.name]</title><style>h3 {margin: 0px; padding: 0px;}</style></head><body><br>
+				<h3>Explosion logs</h3>"}
+	for(var/i in 1 to message_log.len)
+		dat += "Log recording #[i]: [message_log[i]]<br/><br>"
+	dat += "<A href='?src=[REF(src)];delete_log=1'>Delete logs</A><br>"
+	dat += "<hr>"
+	dat += "<A href='?src=[REF(src)];refresh=1'>(Refresh)</A><br>"
+	dat += "</body></html>"
+
+	var/datum/browser/popup = new(user, "computer", name, 400, 500)
+	popup.set_content(dat)

```

Comment:
Join dat list.

Reply:
Where would I put it? I'm just following how every other ui is programmed

Reply:
you just need to switch var/dat to var/list/dat = list()
then dat.Join("") here